### PR TITLE
Convert jsp integration to Instrumenter API

### PR DESCRIPTION
**Pull request by iNikem**

### CODE REVIEW

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentationSingletons.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  private HttpJspPageInstrumentationSingletons() {}
+
+  private static class RenderAttributesExtractor
+      extends AttributesExtractor<HttpServletRequest, Void> {
+
+    @Override
+    protected void onStart(AttributesBuilder attributes, HttpServletRequest request) {
+      if (!CAPTURE_EXPERIMENTAL_SPAN_ATTRIBUTES) {
```
Comment (mateuszrzeszutek) [inline instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentationSingletons.java:59]: Nit: you can just not add the extractor in the instrumenter building phase instead

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/JspCompilationContextInstrumentationSingletons.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        Instrumenter.<JspCompilationContext, Void>newBuilder(
+                GlobalOpenTelemetry.get(),
+                "io.opentelemetry.jsp-2.3",
+                JspCompilationContextInstrumentationSingletons::spanNameOnCompile)
+            .addAttributesExtractor(new CompilationAttributesExtractor())
+            .newInstrumenter(SpanKindExtractor.alwaysInternal());
+  }
+
+  public static String spanNameOnCompile(JspCompilationContext jspCompilationContext) {
+    return jspCompilationContext == null
```
Comment (mateuszrzeszutek) [inline instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/JspCompilationContextInstrumentationSingletons.java:35]: Can `jspCompilationContext` even be null? The instrumenter gets passed the `@Advice.This` instance

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java [hunk 3]
```diff
@@ -52,23 +52,20 @@ public static void onEnter(
         @Advice.Argument(0) HttpServletRequest req,
         @Advice.Local("otelContext") Context context,
         @Advice.Local("otelScope") Scope scope) {
-      context = tracer().startSpan(tracer().spanNameOnRender(req), SpanKind.INTERNAL);
-      tracer().onRender(context, req);
+
+      context = instrumenter().start(Java8BytecodeBridge.currentContext(), req);
```
Comment (mateuszrzeszutek) [inline instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java:56]: Can you add the `shouldStart()` check?

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/JspCompilationContextInstrumentationSingletons.java (hunk 2 above)
Comment (iNikem) [reply instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/JspCompilationContextInstrumentationSingletons.java:35]: It cannot, missed this place

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java (hunk 3 above)
Comment (iNikem) [reply instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java:56]: Does it make sense here? When can it return `true` for INTERNAL spans?

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentationSingletons.java (hunk 1 above)
Comment (iNikem) [reply instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentationSingletons.java:59]: I cannot decide which way is better. Adding extractors conditionally breaks the fluent interface of `InstrumenterBuilder`

File: instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/jsp-2.3/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jsp/HttpJspPageInstrumentation.java:56]: I think we should do that for "future compatibility" (and uniformity in all instrumentations) - if we ever introduce some more elaborate suppresion rules, or have a `shouldStart()` customization interface, etc. There's also the `setDisabled` method.

**Review by mateuszrzeszutek — APPROVED** [review]

**Review by laurit — APPROVED** [review]
Out of cope for this pr but this instrumentation is a bit strange. Really there are 2 instrumentations one for jsp api and another for jasper (jsp engine used on tomcat and many other servers) which probably should be split. For jsp api there should be a separate instrumentation for jakarta api.

**Pull request merged**